Simplify registry API

denis-tingajkin opened the pull request:
Signed-off-by: denis-tingajkin <denis.tingajkin@xored.com>

## Motivation

https://github.com/networkservicemesh/api/issues/39

haiodo commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,43 @@ message Destination {
-    string state = 4;
-}
-
 message NetworkServiceEndpoint {
```
NetworkServiceKind is missing here.

denis-tingajkin replied:
Not sure that it is required for structure NetworkServiceEndpoint, because we can easily handle search case by kind in discovery service.

haiodo replied:
ok, how you plan to register Endpoint or manager?

denis-tingajkin replied:
I would suggest using some network service name convention. For example for forwarders something like `xconn-network-service`, for managers `nsmgr-network-service`.

denis-tingajkin replied:
I also don't mind to add 'kind' field for `NetworkServiceEndpoint`, but I would try to minimalize field count :)

edwarnicke replied:
@haiodo What do you mean by NetworkServiceKind ?

edwarnicke replied:
I'm a bit nervous about baking the 'Kind's in like that into the API.  NSMgrs are a concept in K8s... they might not be in the same way in a different environment (same with Forwarders).  The 'Kinds' listed:

* CROSS_CONNECT_NSE - feels like its just a Network Service, and could be handled as such.
* NETWORK_SERVICE_MANAGER - when/how do you envision us registering an NSMgr?  What Network Service would we register it as providing?
* NETWORK_SERVICE_ENDPOINT

I'm trying to wrap my head around the ways you think we would use these :)

denis-tingajkin replied:
@edwarnicke NOTE: Kind currently used only in discovery service. 

Do you have any idea how we can discover only network service endpoints(i.e. we need to exclude managers, cross-connect nses from response entries)

edwarnicke replied:
I'll turn that question around:

Give me a query that you are concerned would returns a NSMgr or a Cross-Connect NSE incorrectly?

haiodo commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,43 @@ message Destination {
-
 service NetworkServiceRegistry {
-    rpc RegisterNSE (NSERegistration) returns (NSERegistration);
-    rpc BulkRegisterNSE (stream NSERegistration) returns (stream NSERegistration);
```
Did we really need a bulk changes?

denis-tingajkin replied:
This change based on discussion https://docs.google.com/document/d/1JCeBn4QoEl0Nh7Krq9JryFcs-qSX66DGDoewilSv4go/edit?disco=AAAAGjgZrmU

edwarnicke replied:
@haiodo I share your general dislike of BulkChange here.  The reality is that the difference these two is pretty marginal (non-zero, you would still need multiple HTTP/2 streams for multiple RegisterNSE calls... but that overhead is super super small ... its still the same TCP connection).

haiodo replied:
Ok, sounds good.

haiodo commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,43 @@ message Destination {
-}
+message NetworkServiceSelector {
+    NetworkServiceKind kind = 1;
+    string network_service_endpoint_name = 2;
```
you miss network_service_name,

denis-tingajkin replied:
thanks, fixed.

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,44 @@ message Destination {
+    rpc Register (NSERegistration) returns (NSERegistration);
+    rpc Monitor (NetworkServiceSelector) returns (stream NSERegistration);
+    rpc Remove (NetworkServiceSelector) returns (google.protobuf.Empty);
+    rpc Update (NSERegistration) returns (NSERegistration);
```
Why a separate Update method?

haiodo replied:
I've also think we just need a Register/Monitor/Remove

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,44 @@ message Destination {
 
-message NetworkServiceEndpointList {
-    repeated NetworkServiceEndpoint network_service_endpoints = 1;
+enum NetworkServiceKind {
```
Why do we need Kind here?

denis-tingajkin replied:
https://github.com/networkservicemesh/api/pull/41#discussion_r434697556

edwarnicke commented:
I think in general we have two kinds of things (ie, GRPC objects):

1. NetworkServiceEndpoints
2. NetworkServices

For each of those we need to be able to:

* Register
* UnRegister
* Some kind of updating (which may roll into Register or may be distinct)

We also need to be able to do two kinds (minimally) of querying:
* Find - single discrete query
* Monitor - continuous streams of updating

Note: I go back in forth in terms of whether Find/Monitor need to be distinct from each other or can be combined.  Both kinds of queries need various capacity to filter.

denis-tingajkin commented:
@edwarnicke we have a bit discussed with @haiodo your suggestion and we found that we can get rid of registering NSM as an endpoint. On Endpoint registration, NSMgr will simply add its URL to Endpoint.

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,41 @@ message Destination {
-    repeated NetworkServiceEndpoint network_service_endpoints = 4;
+    string url = 6;
+    // Network service manager URL for the endpoint.
+    string network_service_manger_url = 7;
```
Why do we have both url and network_service_manger_url ?

denis-tingajkin replied:
`URL` is a field that is required for callback API goals.
> Endpoint callback address, could be any of unix/tcp/callback schema ones.

`network_service_manger_url` is URL to the network service manager 
who is responsible for this NSE

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,41 @@ message Destination {
+service NetworkServiceEndpointRegistry {
+    rpc RegisterOrUpdate (NetworkServiceEndpoint) returns (NetworkServiceEndpoint);
+    rpc Monitor (NetworkServiceEndpointSelector) returns (stream NetworkServiceEndpoint);
+    rpc Find (NetworkServiceEndpointSelector) returns (stream NetworkServiceEndpoint);
```
What is the difference between Monitor and Find here? :)

denis-tingajkin replied:
In my mind, `Monitor` will be implemented as a `while(true)` goroutine that will notify about changes. 
`Find` should just send all known entries and close the stream.

edwarnicke commented:
Quick question... why are the selectors distinct from the objects?

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,41 @@ message Destination {
-
 message NetworkServiceEndpoint {
     string name = 1;
     string payload = 2;
```
Do we need payload here?

denis-tingajkin replied:
Yes, you are right it is not needed, will remove

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,41 @@ message Destination {
-    string state = 6;
-
+    map<string, string> labels = 4;
+    string state = 5;
```
What are we doing with State?

denis-tingajkin replied:
In monorepo this field used for keeping the state of NSE POD. Looks like we need to remove this field because we going to the independence of k8s

edwarnicke commented:
Let's talk through some use cases, just to get a sense of things.

## UseCase: NSE Registration
1.  NSE calls NetworkServiceEndpointRegistry.RegisterOrUpdate(NetworkServiceEndpoint) providing my name, network service
2. NSMgr catches that call, fills in url, expirationTime, etc, then calls the Registry

(Note: Suppressing the question of who registers the Network Service here)

So far so good.  Questions

1.  What if an NSE wants to register itself with more than one Network Service?
2. What if an NSE wants to register itself with the same Network Service with different label sets?

edwarnicke commented:
## Use Case: NSE Registration Renewal

1.  NSE notes that we are 1/3 of the way from the time it registered to the expirationTime
2.  NSE recalls NetworkServiceEndpointRegistry.RegisterOrUpdate using the Registration returned to it fs initial registration with an updated expireTime.

edwarnicke commented:
# Use Case: Client requests Network Service

1.  Client Requests Network Service
2. NSMgr calls NetworkServiceRegistry.Find and gets back the Network Service
3. NSMgr calls NetworkServiceEndpointRegistry.Find and gets back a list of NSEs
4. NSMgr makes a selection decision
5. ... 

Question:
1.  Are we concerned about having to make two Find calls here?

denis-tingajkin commented:
> ## Use Case: NSE Registration disappears
> Do we want the NSE to call Monitor to trigger fast reregistration if its Registration unexpectedly disappears (ie, Registry dies, NSMgr dies, etc)?

Looks like it will possible with this redesign, but I think we can think about healing cases later when nsmgr will be completed.

denis-tingajkin commented:
> Let's talk through some use cases, just to get a sense of things.
> 
> ## UseCase: NSE Registration
> 1. NSE calls NetworkServiceEndpointRegistry.RegisterOrUpdate(NetworkServiceEndpoint) providing my name, network service
> 2. NSMgr catches that call, fills in url, expirationTime, etc, then calls the Registry
> 
> (Note: Suppressing the question of who registers the Network Service here)
> 
> So far so good. Questions
> 
> 1. What if an NSE wants to register itself with more than one Network Service?
> 2. What if an NSE wants to register itself with the same Network Service with different label sets?

1. Fixed with https://github.com/networkservicemesh/api/pull/41/commits/e9bad6deaada9484afea8ebf1e08590688f13890
2. I guess NSE will just use NetworkServiceEndpointRegistry.RegisterOrUpdate API to update labels set. Do I miss something?

denis-tingajkin commented:
> Quick question... why are the selectors distinct from the objects?

If you mean by objects {NetwrorkService, NetwrorkServiceEndpoint} then selectors can include additional fields in the future, that we will not want to add to _objects_.

edwarnicke commented:
> > Quick question... why are the selectors distinct from the objects?
> 
> If you mean by objects {NetwrorkService, NetwrorkServiceEndpoint} then selectors can include additional fields in the future, that we will not want to add to _objects_.

That makes a lot of sense... but might it make sense to do something like:
```proto
message NetworkServiceEndpointQuery {
     network_service_endpoint = 1;
     ...
}
```

edwarnicke commented:
> > ## Use Case: NSE Registration disappears
> > Do we want the NSE to call Monitor to trigger fast reregistration if its Registration unexpectedly disappears (ie, Registry dies, NSMgr dies, etc)?
> 
> Looks like it will possible with this redesign, but I think we can think about healing cases later when nsmgr will be completed.

Yeah... I'm not saying we have to solve it completely here... just think about it a bit to convince ourselves we can solve it elegantly in the future :)

edwarnicke commented:
> > Let's talk through some use cases, just to get a sense of things.
> > ## UseCase: NSE Registration
> > 
> > 1. NSE calls NetworkServiceEndpointRegistry.RegisterOrUpdate(NetworkServiceEndpoint) providing my name, network service
> > 2. NSMgr catches that call, fills in url, expirationTime, etc, then calls the Registry
> > 
> > (Note: Suppressing the question of who registers the Network Service here)
> > So far so good. Questions
> > 
> > 1. What if an NSE wants to register itself with more than one Network Service?
> > 2. What if an NSE wants to register itself with the same Network Service with different label sets?
> 
> 1. Fixed with [e9bad6d](https://github.com/networkservicemesh/api/commit/e9bad6deaada9484afea8ebf1e08590688f13890)
> 2. I guess NSE will just use NetworkServiceEndpointRegistry.RegisterOrUpdate API to update labels set. Do I miss something?

On (2) I was thinking more like "What happens if a Network Service Endpoint wants to register with *two* simultaenous sets of labels"... I'm not even sure that's a use case we care about... but wanted to ask the question and think about it.

denis-tingajkin commented:
> > > Quick question... why are the selectors distinct from the objects?
> > 
> > 
> > If you mean by objects {NetwrorkService, NetwrorkServiceEndpoint} then selectors can include additional fields in the future, that we will not want to add to _objects_.
> 
> That makes a lot of sense... but might it make sense to do something like:
> 
> ```protobuf
> message NetworkServiceEndpointQuery {
>      network_service_endpoint = 1;
>      ...
> }
> ```

Thanks, fixed

denis-tingajkin commented:
> > > Let's talk through some use cases, just to get a sense of things.
> > > ## UseCase: NSE Registration
> > > 
> > > 1. NSE calls NetworkServiceEndpointRegistry.RegisterOrUpdate(NetworkServiceEndpoint) providing my name, network service
> > > 2. NSMgr catches that call, fills in url, expirationTime, etc, then calls the Registry
> > > 
> > > (Note: Suppressing the question of who registers the Network Service here)
> > > So far so good. Questions
> > > 
> > > 1. What if an NSE wants to register itself with more than one Network Service?
> > > 2. What if an NSE wants to register itself with the same Network Service with different label sets?
> > 
> > 
> > 
> > 1. Fixed with [e9bad6d](https://github.com/networkservicemesh/api/commit/e9bad6deaada9484afea8ebf1e08590688f13890)
> > 2. I guess NSE will just use NetworkServiceEndpointRegistry.RegisterOrUpdate API to update labels set. Do I miss something?
> 
> On (2) I was thinking more like "What happens if a Network Service Endpoint wants to register with _two_ simultaenous sets of labels"... I'm not even sure that's a use case we care about... but wanted to ask the question and think about it.

I think this case makes sense. I've changed labels to
```proto
message NetworkServiceLabels {
    map<string, string> labels = 1;
}
...
    map<string, NetworkServiceLabels> labels = 3;
```

haiodo commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,44 @@ message Destination {
+    map<string, string> labels = 3;
     // Endpoint callback address, could be any of unix/tcp/callback schema ones.
-    string url = 7;
+    string url = 4;
```
Should we consider `repeated string url = 4` so when endpoint is registered from Nsmgr, it will have two urls, one is url of manager and one it internal one?

denis-tingajkin replied:
Hmm.. probably then we should use the map
```
map<string, string>  urls
```
to get rid of problems with URL order convention,
Then we can use it something like
```
nse_url: callback://addr
nsm_url: tcp://addr
extra_url: ....
```
@edwarnicke , @haiodo What do you think?

haiodo replied:
If we have scenario when it could be used map is better structured in compare to list. We we not have such scenario when separate field is ok.

haiodo commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,48 @@ message Destination {
-    map<string, NetworkServiceManager> network_service_managers = 3;
-    repeated NetworkServiceEndpoint network_service_endpoints = 4;
+service NetworkServiceRegistry {
+    rpc RegisterOrUpdate (NetworkService) returns (NetworkService);
```
I think RegisterOrUpdate is same as Register for this context, so why we need so long function name :-)

denis-tingajkin replied:
ok, fixed.

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,48 @@ message Destination {
+service NetworkServiceEndpointRegistry {
+    rpc Register (NetworkServiceEndpoint) returns (NetworkServiceEndpoint);
+    rpc Monitor (NetworkServiceEndpointQuery) returns (stream NetworkServiceEndpoint);
+    rpc Find (NetworkServiceEndpointQuery) returns (stream NetworkServiceEndpointList);
```
We would stream here a Response of some sort, not a List.  The point of streaming is that we don't need a list :)

denis-tingajkin replied:
> The point of streaming is that we don't need a list :)

In general, you are right, but lest consider the case when we need to send a million endpoints. So if we'll use steam of lists where list length is `length = (GRPC_MAX_MESSAGE_SIZE/sizeof(NetworkServiceEndpoint))` then it'll work in length time faster them single message stream.

denis-tingajkin replied:
@edwarnicke make sense?

denis-tingajkin replied:
Finally fixed.

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,48 @@ message Destination {
-    rpc RemoveNSE (RemoveNSERequest) returns (google.protobuf.Empty);
+message NetworkServiceEndpointQuery {
+    NetworkServiceEndpoint network_service_endpoint = 1;
+    NetworkService network_service = 2;
```
Why do we have both a NetworkServiceEndpoint and a NetworkService object in this query?

denis-tingajkin replied:
Related to this comment https://github.com/networkservicemesh/api/pull/41#issuecomment-638912388

fkautz replied:
The QPS for gRPC in go is pretty high.

https://performance-dot-grpc-testing.appspot.com/explore?dashboard=5636470266134528&widget=828849126&container=1808124444

I'd prefer to start without a list and introduce a list if gRPC throughput turns out to be our bottleneck when we perform stress tests.

We also have other techniques we can use to reduce the total quantity of endpoints transferred such as summarization.

denis-tingajkin replied:
@fkautz I've little tested locally(If you are interested I can provide the testing on CI) and got next results:


n = 1e7

Steam of lists where the length of the list is 10000, time :11.556816ms
Steam of single object: time: 30.680219342s

n = 1e9

Steam of lists where the length of the list is 10000, time :14.89991ms
Steam of single object: time: 30.680219342s


n = 1e10

Steam of lists where the length of the list is 10000, time: 46.025932ms
Steam of single object: time: 5m10.775024696s


_n is count of endpoinst.



Also, I've tested the case when used list without stream. It works less than second only with `n=6`, time: 621.380154ms (for n=7 time is 6.172965725s).

denis-tingajkin replied:
So it looks like a stream of list has the best performance.

edwarnicke replied:
> So it looks like a stream + list has the best performance.

True, and interesting... but perf is pretty close up to 10^7... I'm also most curious the *why* under the increasing perf difference... it feels like there is something interesting in the system going on.

I tend to agree with @fkautz , that we should start with the non-list stream case.  I'm pretty sure that by the time you are playing in the 10^7 range, other elements of the system, like the underlying DB being read from will be the dominant factor.

fkautz replied:
My preference is still for a single object rather than a list for simplicity.

In the case of n = 1e7:

Steam of lists where the length of the list is 10000, time :11.556816ms
Steam of single object: time: 22.952515ms

That's a theoretical limit of approximately 450 million messages / second. There are articles from companies like Netflix about how they achieved 1 million inserts / second in a 60 node Cassandra cluster. I think the single object approach is ok in this scenario. It's highly unlikely gRPC will be our bottleneck here. If it is, we can horizontally scale it. Even if you go with an append only queue such as Kafka, you're looking at 1 or 2M appends / second for a cluster of 3.

In short, gRPC will not be our bottleneck. We're looking at database and KV stores as the bottleneck here.

fkautz replied:
That being said, I do worry about latency between systems. It would be good to see what happens when you add 10ms or 100ms of latency to the network. That may provide better context on how we proceed.

denis-tingajkin replied:
> That being said, I do worry about latency between systems. It would be good to see what happens when you add 10ms or 100ms of latency to the network. That may provide better context on how we proceed.

I very like the direction you are thinking here, I think we need to add this kind of stress tests when main NSM components will be completed to see what in the system is our bottleneck.

fkautz commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,40 @@ message Destination {
-    repeated NetworkServiceEndpoint network_service_endpoints = 4;
+    string url = 4;
+    // Network service manager URL for the endpoint.
+    string network_service_manger_url = 5;
```
Do we need url and network_service_manager_url?

Is it possible to get away with using one url or do we need both?

denis-tingajkin replied:
> Do we need url and network_service_manager_url?
Yes, we need both URLs
As I understand: 
`url` is using for callback goals.
`network_service_manger_url` is planned to use for different goals such as connecting to remote nsmgr, discover all endpoints related to the specific manager and etc.

edwarnicke replied:
Yeah... I still don't see why we need two urls.  If you want callback goals (ie, localbypass) ... that's data for use locally inside the K8s NSMgr.  It shouldn't leak out to the registry.

denis-tingajkin replied:
ok, we've little bit discussed this moment with @haiodo and looks like we can remove one of URLs.

denis-tingajkin replied:
Fixed https://github.com/networkservicemesh/api/pull/41/commits/3b53da4e4e1a26185f782b92ce1ca044b516c38a

edwarnicke commented on pkg/api/registry/registry.proto:
```diff
@@ -22,61 +22,40 @@ message Destination {
-}
+message NetworkServiceEndpointQuery {
+    NetworkServiceEndpoint network_service_endpoint = 1;
+    NetworkService network_service = 2;
```
I still don't follow why we have both the NetworkServiceEndpoint and NetworkService objects in the NetworkServiceEndpoint query.  Could you explain?

denis-tingajkin replied:
Sure, is this case make sense? https://github.com/networkservicemesh/api/pull/41#issuecomment-638912388

denis-tingajkin replied:
Fixed https://github.com/networkservicemesh/api/pull/41/commits/3b53da4e4e1a26185f782b92ce1ca044b516c38a

edwarnicke replied:
Ah... you took my question as direction :)

So my thought is this... that anything NetworkService related here we'd need to communicate can be done with the NetworkServiceEndpoint.NetworkService field

denis-tingajkin replied:
@edwarnicke NetworkServiceEndpoint has not field NetworkService. It has only the name of network service

edwarnicke replied:
Yes... but other than the name... what else would we look up a NetworkServiceEndpoint on :)
